piccolo_api/shared/auth/hooks.py

```diff
@@ -0,0 +1,136 @@
+from __future__ import annotations
+
+import dataclasses
+import inspect
+import logging
+import typing as t
+
+from piccolo.apps.user.tables import BaseUser
+
+PreLoginHook = t.Union[
+    t.Callable[[str], t.Optional[str]],
+    t.Callable[[str], t.Awaitable[t.Optional[str]]],
+]
+LoginSuccessHook = t.Union[
+    t.Callable[[BaseUser], t.Optional[str]],
+    t.Callable[[BaseUser], t.Awaitable[t.Optional[str]]],
+]
+LoginFailureHook = t.Union[
+    t.Callable[[str], t.Optional[str]],
+    t.Callable[[str], t.Awaitable[t.Optional[str]]],
+]
+
+
+logger = logging.getLogger(__file__)
+
+
+@dataclasses.dataclass
+class LoginHooks:
+    """
+    Allows you to run custom logic during login. A hook can be a function or
+    coroutine.
+
+    Here's an example using :class:`session_login <piccolo_api.session_auth.endpoints.session_login>`:
+
+    .. code-block:: python
+
+        def check_ban_list(username: str, **kwargs):
+            '''
+            An example pre_login hook.
+            '''
+            if username in ('nuisance@pest.com',):
```

Comment:
In this example, I think you should pass tuple of usernames not emails because this always return `False` and doesn't trigger this hook. E.g `if username in ("john", "bill",):`

Reply:
Good point - I'll change that.